### ISSUE TITLE
User ID in jenkins optionally UPN

### ISSUE THREAD

**Issue by davidhonig** [jenkinsci/azure-ad-plugin#16]
Hi, for granular permissions the matrix authentication is not enough. It would be brilliant if there is configurable what is User ID in jenkins. User ID 67896789-6789-6789789-6789789789 is quite unusable in Role Strategy plugin. And for some reasons it would be great if this is UPN, especially when migrating stuff it is easy to pair e-mail adresses/upns but not ID's. It would be great to be able to use upn without the @xxxxxxx.xxx part too.

**Comment by Yatheesh-Rai** [comment]
Hi,
For us using this plugin , MFA and SSO including the authentication and authorization part worked without any issue.
However the issue raised above by davidhonig  is the major blocker.
Could you please enhance the plugin to use UPN as user short name from AD instead of User object ID.

**Comment by Yatheesh-Rai** [comment]
Hi gavinfish,

Thanks for developing this plugin, this resolves most of our test cases for authentication and authorization  to work in jenkins.
however the user id (object ID) - (a5e5cdad-c868-4073-9fe3-739dfa75fcef) which is unusable for us in the authorization strategy/token generation. We are looking for it to come as User short name (yrai) ideally the  AD sAMAccountName or userPrincipalName -

example - Yatheesh Rai is full username , so the user id will be  in AD is "yrai"
 It would be great to be able to use upn without the @xxxxxxx.xxx part too.

 
I know you have initiated the plugin enhancement. Since we are waiting for this feature to available in plugin and we are in a tight schedule to role out this plugin to production, please let me know when can we expect the delivery of enhanced feature. It would be great help, if you can make it to release asap 

Thanks again,
Yatheesh

**Comment by gavinfish** [comment]
hi @Yatheesh-Rai ,
I have made a PR for it. I will make a release when the PR merged. Sorry for the latency.

**Comment by Yatheesh-Rai** [comment]
Thanks @gavinfish for enhancing the plugin 
We are looking for user nickname/shortname (SAMAccountName) instead of username
An example -  Yatheesh Rai is my username and Yatheesh.Rai@xxx.com is my UPN, but my SAMAccountName is "yrai".

We are expecting the SAMAccountName to appear like below
https://JenkinsURL/user/yrai/configure

Attached the screenshot of how we want it to be.
<img width="998" alt="screen shot 2018-09-18 at 11 30 11 am" src="https://user-images.githubusercontent.com/20515317/45667461-e0444a00-bb36-11e8-9b82-8948aa8edda1.png">

Please check if is feasible 

Thanks
Yatheesh

**Comment by gavinfish** [comment]
Hi @Yatheesh-Rai , is SAMAccountName unique? As discussed [here](https://github.com/jenkinsci/azure-ad-plugin/pull/19#discussion_r218290829), the Jenkins user id should be unique.

**Comment by davidhonig** [comment]
Hi, guys,

No SAMAccont name is not unique in MS infrastructure, you need to use UPN for that.

18. 9. 2018 v 8:31, Jie Shen <notifications@github.com<mailto:notifications@github.com>>:


Hi @Yatheesh-Rai<https://github.com/Yatheesh-Rai> , is SAMAccountName unique? As discussed here<https://github.com/jenkinsci/azure-ad-plugin/pull/19#discussion_r218290829>, the Jenkins user id should be unique.

**Comment by gavinfish** [comment]
Agree with @davidhonig , I think we should use UPN or mail notaion. Personally I will choose mail notation.

**Comment by davidhonig** [comment]
Yep UPN in mail notation ;-)

**Comment by Yatheesh-Rai** [comment]
Hi @gavinfish 

I agree UPN is better in most of the case. But in our case, we have a need to map sAMAccountName as Jenkins user ID.

Our AD server is on a single domain. Since we retrieve user details from single domain , we are fine to use sAMAccountName. 
Could you please let us know how we can configure the sAMAccountName claim in Azure AD and  the claim used for retrieving the  sAMAccountName in Azure AD plugin.

**Comment by davidhonig** [comment]
Samaccount name is not pure Azure AD parameter, it is synced from on-prem. So that will work only on hybrid envs. But I understand that if you map mail to UPN not samaccount name to UPN that you want to use SAMA... for future is better teach users use mail even on prem...

**Comment by Yatheesh-Rai** [comment]
We are looking for sAMAccountName for a specific purpose in Jenkins.
Please provide a way to set up the claim for sAMAccountName in Azure AD, and how to retrieve the value of it using the Azure AD plugin.

**Comment by gavinfish** [comment]
> We are looking for sAMAccountName for a specific purpose in Jenkins.
> Please provide a way to set up the claim for sAMAccountName in Azure AD, and how to retrieve the value of it using the Azure AD plugin.

@LGDoor Do you have any suggestions?

**Comment by LGDoor** [comment]
No idea about how to configure the sAMAccountName in Azure AD. It's most likely not available.

**Comment by Yatheesh-Rai** [comment]
@gavinfish @LGDoor 
Thanks for the response. I am quiet new to Azure AD claims 
We have a specific requirement in Jenkins for which we would need sAMAccountName

Does any of the below will help in retrieving the sAMAccountName

"onpremisessamaccountname" 
"windowsaccountname"
"nameidentifier" 
"nickname"

https://docs.microsoft.com/en-us/azure/active-directory/active-directory-claims-mapping

**Comment by Yatheesh-Rai** [comment]
Hi @gavinfish 

With the enhanced plugin we found a defect where in even if I am part of Authorization strategy , I will not be able to access the jenkins URL using curl or  consume the REST API 

I could see missing permission for the user in curl console output

I am added as below in authorization strategy
Yatheesh Rai (a5e5cdad-c868-4073-9fe3-739dfa75fcef)

curl -k -u 'Yatheesh Rai:jenkinsapitoken' 'https://JenkinsURL/jenkins/user/Yatheesh%20Rai/configure'

FYI, the same was working before the plugin enhancement - when it was using the objectID as below, we were able to access jenkins url using curl and the REST APIs  

curl -k -u 'a5e5cdad-c868-4073-9fe3-739dfa75fcef:jenkinsapitoken' 'https://JenkinsURL/jenkins/user/a5e5cdad-c868-4073-9fe3-739dfa75fcef/configure'

Please fix this issue

Thanks
Yatheesh

**Comment by gavinfish** [comment]
@Yatheesh-Rai I have tried to curl the configuration and it works for me. Did you generate a new token for the new user with id `Yatheesh Rai`? Since there should be two Jenkins users in your server with the same object id and different user id. They manage their own tokens.

**Comment by Yatheesh-Rai** [comment]
Hi @gavinfish 

Thanks for helping us !
Sorry for the back to back questions.
It is not working for me still.

curl -k -u 'Yatheesh Rai:APIToken' 'https://JenkinsURL/jenkins/users/Yatheesh%20Rai/configure'

curl -X POST -u 'Yatheesh Rai:APItoken' https://JenkinsURL/jenkins/job/_playground/job/Yatheesh/job/apitoken


I have removed the old references of users created in my name

Now I have only "yatheesh rai" in jenkins users

I tried regenerating the token for "yatheesh rai" 

The error that I see on the curl output is;
<h1>Access Denied</h1><p class="error">yatheesh rai is missing the Overall/Read permission

I have added to Azure Active Directory Matrix-based security with overall permission 

Yatheesh Rai (a5e5cdad-c868-566f-re34-739dfa75fcef) 

<img width="1159" alt="screen shot 2018-09-20 at 11 16 03 am" src="https://user-images.githubusercontent.com/20515317/45798296-fed94b00-bcc6-11e8-9a63-c79878a29840.png">


My user id is;
<img width="859" alt="screen shot 2018-09-20 at 11 11 55 am" src="https://user-images.githubusercontent.com/20515317/45798068-0e0bc900-bcc6-11e8-980e-b80da4fd10f5.png">

Can you please help me on what I'm doing wrong 

Thanks again for you help !

Thanks
Yatheesh

**Comment by gavinfish** [comment]
@Yatheesh-Rai I just notice you use `Yatheesh Rai` as Jenkins id? The newest plugin should use UPN as the Jenkins id. So please check your Jenkins user id again. You can find it at "https://jenkinsURL/asynchPeople/"

**Comment by Yatheesh-Rai** [comment]
Hi @gavinfish 

Sorry to trouble you with this issue.
Could you please let me know did you get a chance to check this.
We can set up a call to have a debugging session to understand more about the issue that we are facing.


Thanks for your time

**Comment by gavinfish** [comment]
Hi @Yatheesh-Rai ,
Sorry for the late response since I was on vocation. I agree that use `Yatheesh Rai (yatheesh.rai@logmein.com) in Azure Active Directory Matrix-based security` will be more reasonable now. I will make a PR later.

**Comment by Yatheesh-Rai** [comment]
@gavinfish 
It is the blocker for us.
When we add Yatheesh.rai@logmein.com to authorization matrix , REST API call is working but Jenkins Permission is missing
<img width="498" alt="screen shot 2018-09-25 at 11 59 58 am" src="https://user-images.githubusercontent.com/20515317/45996660-9ae5c680-c0ba-11e8-84ba-f1848ede80ae.png">


If we add Yatheesh Rai (a5e5cdad-c868-4073-9fe3-739dfa75fcef) to authorization matrix (this is the default option to add users in current plugin version), Jenkins access permission is working , but REST API call is not working.
<img width="479" alt="screen shot 2018-09-25 at 12 00 08 pm" src="https://user-images.githubusercontent.com/20515317/45996672-a5a05b80-c0ba-11e8-8c08-cadde9eb7c24.png">

If you want we can have a call to get on the same page.

**Comment by gavinfish** [comment]
@Yatheesh-Rai 
If you add `Yatheesh Rai (a5e5cdad-c868-4073-9fe3-739dfa75fcef)` to authorization matrix, you need to find out what it the Jenkins user id(UPN now) of the user for REST API. There should be two Jenkins user with the same name in your jenkins server now. But the first one's id is `a5e5cdad-c868-4073-9fe3-739dfa75fcef` and the other one is the UPN. Please make sure you use the right Jenkins user id and its related token to call the API. I can succeed to call the API with the second setting now.

**Comment by Yatheesh-Rai** [comment]
Hi @gavinfish 

We have tried with the new user who is accessing Jenkins for the first time.
For him the REST API call is not working using the UPN as USER ID.

May be we are understanding it wrong. 
Can we have a quick call to know this better.

Thanks
Yatheesh

**Comment by gavinfish** [comment]
@Yatheesh-Rai , my working e-mail is jieshe@microsoft.com, contact me for further discussion, Thanks.

**Comment by davidhonig** [comment]
Hi, just confirming that user gets a new error message where UPN is instead of ID - that is brilliant, but If I add UPN instead of the ID number in the authorisation matrix the rights are not added for the user. ID still works OK

**Comment by shantur** [comment]
@gavinfish : https://github.com/jenkinsci/azure-ad-plugin/pull/39 should fix it without breaking older configuration

**Comment by opack** [comment]
Hello guys!
We have the same problem: hard to manage user permissions with object-IDs :-( It is a blocker for us, and we are thinking of using the simple SAML plugin instead, which let's us choose the field to use for ID.
By any chance, do you have some estimate on the next release packaging the feature discussed in this thread? Just to know if I can stole a little...

**Comment by shantur** [comment]
@opack : You can try #39. It fixes the issue without breaking older configurations.

**Comment by opack** [comment]
Hi @shantur , thanks for the answer :-) But I already saw this but have no mean to retrieve / rebuild the `azure-ad.jar` containing the modification :-( Could you help me with this please? 👼  I think anyone coming here would be delighted to have this fixed jar before it is delivered in the next release 😃

**Comment by shantur** [comment]
Try : https://quickfileshare.org/1LvM/azure-ad.hpi.zip

**Comment by davidhonig** [comment]
@shantur this works like a charm! I'm looking forward to be approved!!!

**Comment by opack** [comment]
Excuse me, but now... this is WAY TOO COOL ! 😄 Thanks a lot @shantur it is awesome 👍 ! Now I have manageable user names !

**Comment by saitotqr** [comment]
I have an problem when invoking Jenkins-API with user-token when using Jenkins-Azure-AD Plugin. Same of this issue. 
https://issues.jenkins-ci.org/browse/JENKINS-54115

Im using Azure Active Directory Matrix based security. I added Azure AD Group, and invoke API with user-token, but remote api doesn't work. But when I added user (like john.smith) can invoke.

Actually, I want to add&remove users only in Azure-AD group. so this is inconveniense.
